### PR TITLE
Permit attributed variables inside signatured subs (Fixes GH #21158)

### LINKED ISSUE #21158: field syntax sometimes provokes "subroutine attributes must come before the signature" error
This is fine:

```
$ perl5.38.0-RC1 -Mv5.38 -Mexperimental=class -c
class Bar {
  field $x :param;
}
- syntax OK
```

This too:

```
$ perl5.38.0-RC1 -Mv5.38 -Mexperimental=class -c
{ sub foo($x,$y) { } }

class Bar {
  field $x :param;
}
- syntax OK
```

However, without the block wrapping:

```
$ perl5.38.0-RC1 -Mv5.38 -Mexperimental=class -c
sub foo($x,$y) { }

class Bar {
  field $x :param;
Subroutine attributes must come before the signature at - line 4.
```

### PR DESCRIPTION
**Pull request by leonerd**
Fixes #21158

It's been broken since at least 5.36 and probably older, so I don't feel there's any rush to get this in before 5.38.0, but if folks are happy I guess it could go in.

### CODE REVIEW

File: t/op/signatures.t [hunk 1]
```diff
@@ -1713,6 +1713,14 @@ while(<$kh>) {
     is($warnings, "", 'No warnings emitted within scope of  no warnings "experimental"');
 }
 
+# GH #21158
+#   The :baz attribute is unrecognised but in the current implementation that
+#   is only checked at runtime, and we never invoke the function so this
+#   should be fine.
+ok(defined eval 'sub gt21158 ($x) { my $bar :baz; } "ok"',
```
Comment (ilmari) [inline t/op/signatures.t]: Shouldn't that be `gh21158`?

**Review by ilmari — CHANGES_REQUESTED** [review]

File: t/op/signatures.t (hunk 1 above)
Comment (leonerd) [reply t/op/signatures.t]: Ahyes. Force-pushed.

**Comment by rwp0** [comment]
Thanks.

May also consider putting this in the PR description: Fixes #21158 (and not just on title)

So the PR is linked to issue properly by GitHub.

image

**Comment by leonerd** [comment]
> May also consider putting this in the PR description: Fixes #21158 (and not just on title)

Oh huh. I didn't realise that made a difference.

**Comment by rwp0** [comment]
> > May also consider putting this in the PR description: Fixes #21158 (and not just on title)
> 
> Oh huh. I didn't realise that made a difference.

Thanks, now the issue also appears here in the PR list:

image

**Comment by perigrin** [comment]
My only argument with having this possibly go in with 5.38.0-RC2 is that before we only ever really use non-subroutine attributes (in core) with shared variables so it's fairly uncommon. With `feature 'class'` they're more common as they're the way to add arguments to the class constructor.

Given that `feature 'class'` is experimental though, I don't know that it is compelling at this late stage.

**Comment by leonerd** [comment]
@perigrin Indeed. It seems fairly rare to encounter it up until now. Perhaps best to wait until 5.39.x then. Shall I tag it defer-next-dev?

**Comment by tonycoz** [comment]
This might also be a fix for #19245

**Review by rwp0 — APPROVED** [review]

**Comment by perigrin** [comment]
@leonerd I think that's up to people higher up the p5p food chain than me. I've re-written the demo code I had to avoid the issue so I'm good for the foreseeable future. If 5.40 is the right answer then that's the right answer.

**Review by ilmari — APPROVED** [review]

**Review by rwp0 — APPROVED** [review]

**Comment by rwp0** [comment]
Merge in for 5.38.0 🐫

**Review by esabol — APPROVED** [review]

**Comment by jkeenan** [comment]
> @perigrin Indeed. It seems fairly rare to encounter it up until now. Perhaps best to wait until 5.39.x then. Shall I tag it defer-next-dev?

@leonerd, I've removed the 'defer-next-dev' label.  Is this p.r. ready to merge?

**Comment by esabol** [comment]
> > @perigrin Indeed. It seems fairly rare to encounter it up until now. Perhaps best to wait until 5.39.x then. Shall I tag it defer-next-dev?
> 
> @leonerd, I've removed the 'defer-next-dev' label. Is this p.r. ready to merge?

One of the "minitests" failed in what seems to be a non-spurious way:

>> Failed test 1245 - Signatured subroutine permits attributed scalar at op/signatures.t line 1720
>> Error was Can't locate attributes.pm in INC (you may need to install the attributes module) INC entries checked: ../lib /home/runner/work/perl5/perl5/cpan/AutoLoader/lib /home/runner/work/perl5/perl5/dist/Carp/lib /home/runner/work/perl5/perl5/dist/PathTools /home/runner/work/perl5/perl5/dist/PathTools/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Install/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-MakeMaker/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Manifest/lib /home/runner/work/perl5/perl5/cpan/File-Path/lib /home/runner/work/perl5/perl5/ext/re /home/runner/work/perl5/perl5/dist/Term-ReadLine/lib /home/runner/work/perl5/perl5/dist/Exporter/lib /home/runner/work/perl5/perl5/ext/File-Find/lib /home/runner/work/perl5/perl5/cpan/Text-Tabs/lib /home/runner/work/perl5/perl5/dist/constant/lib /home/runner/work/perl5/perl5/cpan/version/lib /home/runner/work/perl5/perl5/cpan/Getopt-Long/lib /home/runner/work/perl5/perl5/cpan/Text-ParseWords/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-PL2Bat/lib /home/runner/work/perl5/perl5/lib .. . .) at (eval 828) line 1, <$kh> line 382.

**Comment by jkeenan** [comment]
> > > @perigrin Indeed. It seems fairly rare to encounter it up until now. Perhaps best to wait until 5.39.x then. Shall I tag it defer-next-dev?
> > 
> > 
> > @leonerd, I've removed the 'defer-next-dev' label. Is this p.r. ready to merge?
> 
> One of the "minitests" failed in what seems to be a non-spurious way:
> 
> > > Failed test 1245 - Signatured subroutine permits attributed scalar at op/signatures.t line 1720
> > > Error was Can't locate attributes.pm in INC (you may need to install the attributes module) INC entries checked: ../lib /home/runner/work/perl5/perl5/cpan/AutoLoader/lib /home/runner/work/perl5/perl5/dist/Carp/lib /home/runner/work/perl5/perl5/dist/PathTools /home/runner/work/perl5/perl5/dist/PathTools/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Install/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-MakeMaker/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Manifest/lib /home/runner/work/perl5/perl5/cpan/File-Path/lib /home/runner/work/perl5/perl5/ext/re /home/runner/work/perl5/perl5/dist/Term-ReadLine/lib /home/runner/work/perl5/perl5/dist/Exporter/lib /home/runner/work/perl5/perl5/ext/File-Find/lib /home/runner/work/perl5/perl5/cpan/Text-Tabs/lib /home/runner/work/perl5/perl5/dist/constant/lib /home/runner/work/perl5/perl5/cpan/version/lib /home/runner/work/perl5/perl5/cpan/Getopt-Long/lib /home/runner/work/perl5/perl5/cpan/Text-ParseWords/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-PL2Bat/lib /home/runner/work/perl5/perl5/lib .. . .) at (eval 828) line 1, <$kh> line 382.

An excellent catch!  @leonerd, can you revise the p.r. to skip tests as needed -- probably something like one of these:
```
$ ack miniperl t/op | ack attributes
t/op/getpid.t:10:	"no dynamic loading on miniperl, no threads/attributes"
t/op/state.t:174:skip "no attributes in miniperl", 3, if is_miniperl;
t/op/state.t:395:        skip_if_miniperl("miniperl can't load attributes.pm", 1)
t/op/attrs.t:9:    skip_all_if_miniperl("miniperl can't load attributes");
t/op/attrhand.t:7:    skip_all_if_miniperl("miniperl can't load attributes");
t/op/decl-refs.t:69:            skip_if_miniperl "No attributes on miniperl", 2;
t/op/decl-refs.t:92:            skip_if_miniperl "No attributes on miniperl", 2;
t/op/attrproto.t:10:    skip_all_if_miniperl("miniperl can't load attributes");
```

**Comment by jkeenan** [comment]
> > > > @perigrin Indeed. It seems fairly rare to encounter it up until now. Perhaps best to wait until 5.39.x then. Shall I tag it defer-next-dev?
> > > 
> > > 
> > > @leonerd, I've removed the 'defer-next-dev' label. Is this p.r. ready to merge?
> > 
> > 
> > One of the "minitests" failed in what seems to be a non-spurious way:
> > > > Failed test 1245 - Signatured subroutine permits attributed scalar at op/signatures.t line 1720
> > > > Error was Can't locate attributes.pm in INC (you may need to install the attributes module) INC entries checked: ../lib /home/runner/work/perl5/perl5/cpan/AutoLoader/lib /home/runner/work/perl5/perl5/dist/Carp/lib /home/runner/work/perl5/perl5/dist/PathTools /home/runner/work/perl5/perl5/dist/PathTools/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Install/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-MakeMaker/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-Manifest/lib /home/runner/work/perl5/perl5/cpan/File-Path/lib /home/runner/work/perl5/perl5/ext/re /home/runner/work/perl5/perl5/dist/Term-ReadLine/lib /home/runner/work/perl5/perl5/dist/Exporter/lib /home/runner/work/perl5/perl5/ext/File-Find/lib /home/runner/work/perl5/perl5/cpan/Text-Tabs/lib /home/runner/work/perl5/perl5/dist/constant/lib /home/runner/work/perl5/perl5/cpan/version/lib /home/runner/work/perl5/perl5/cpan/Getopt-Long/lib /home/runner/work/perl5/perl5/cpan/Text-ParseWords/lib /home/runner/work/perl5/perl5/cpan/ExtUtils-PL2Bat/lib /home/runner/work/perl5/perl5/lib .. . .) at (eval 828) line 1, <$kh> line 382.
> 
> An excellent catch! @leonerd, can you revise the p.r. to skip tests as needed -- probably something like one of these:
> 
> ```
> $ ack miniperl t/op | ack attributes
> t/op/getpid.t:10:	"no dynamic loading on miniperl, no threads/attributes"
> t/op/state.t:174:skip "no attributes in miniperl", 3, if is_miniperl;
> t/op/state.t:395:        skip_if_miniperl("miniperl can't load attributes.pm", 1)
> t/op/attrs.t:9:    skip_all_if_miniperl("miniperl can't load attributes");
> t/op/attrhand.t:7:    skip_all_if_miniperl("miniperl can't load attributes");
> t/op/decl-refs.t:69:            skip_if_miniperl "No attributes on miniperl", 2;
> t/op/decl-refs.t:92:            skip_if_miniperl "No attributes on miniperl", 2;
> t/op/attrproto.t:10:    skip_all_if_miniperl("miniperl can't load attributes");
> ```

@leonerd please see above suggestion for revision.

**Comment by leonerd** [comment]
Repushed; will await CI

**Review by rwp0 — APPROVED** [review]

**Review by esabol — APPROVED** [review]

**Pull request merged**